Enable the Ansible task to optionally cleanup after itself.

andrewschoen opened the pull request:
If given the config value of ``cleanup`` the ansible task will rerun the
same playbook during the teardown stage with a ``teardown`` var added
with a value of True.  This will allow the playbook to cleanup after
itself during teardown, if the playbook supports this feature.

The tests pass, but I haven't actually ran a job with this yet.

ceph-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.ceph.com//job/teuthology-pull-requests/1527/
Tests passed for this pull request.

andrewschoen commented:
I just tested this manually, works great.

Relevant portion of the logs:

```
2015-06-18 15:08:11,568.568 INFO:teuthology.task.ansible:Playbook: [{'hosts': 'testnodes', 'roles': ['downstream-setup']}]
2015-06-18 15:08:11,569.569 DEBUG:teuthology.task.ansible:Running ansible-playbook -v --extra-vars '{"yum_repos": [{"url": "http://puddle.ceph.redhat.com/puddles/RHCeph/1.3-RHEL-7/l
atest/Server-RH7-CEPH-MON-1.3.repo", "name": "mon"}], "ansible_ssh_user": "ubuntu"}' -i /etc/ansible/hosts --limit magna085.ceph.redhat.com /home/andrewschoen/src/ceph-cm-ansible_ma
ster/downstream_setup.yml
2015-06-18 15:08:12,437.437 INFO:teuthology.task.ansible.out:
PLAY [testnodes] **************************************************************

2015-06-18 15:08:12,439.439 INFO:teuthology.task.ansible.out:
GATHERING FACTS ***************************************************************

2015-06-18 15:08:13,961.961 INFO:teuthology.task.ansible.out:ok: [magna085.ceph.redhat.com]

2015-06-18 15:08:13,963.963 INFO:teuthology.task.ansible.out:
TASK: [downstream-setup | Download yum repos to /etc/yum.repos.d] *************

2015-06-18 15:08:14,136.136 INFO:teuthology.task.ansible.out:changed: [magna085.ceph.redhat.com] => (item={u'url': u'http://puddle.ceph.redhat.com/puddles/RHCeph/1.3-RHEL-7/latest/S
erver-RH7-CEPH-MON-1.3.repo', u'name': u'mon'}) => {"changed": true, "checksum": "0aea7f8ff8c917f4d0bfebbbbbc06c86f310e63c", "dest": "/etc/yum.repos.d/mon.repo", "gid": 0, "group":
"root", "item": {"name": "mon", "url": "http://puddle.ceph.redhat.com/puddles/RHCeph/1.3-RHEL-7/latest/Server-RH7-CEPH-MON-1.3.repo"}, "md5sum": "6f8c29b8bb5b629734f909eb7d1e2c05",
"mode": "0644", "msg": "OK (608 bytes)", "owner": "root", "sha256sum": "", "size": 608, "src": "/tmp/tmpiBsJY4", "state": "file", "uid": 0, "url": "http://puddle.ceph.redhat.com/pud
dles/RHCeph/1.3-RHEL-7/latest/Server-RH7-CEPH-MON-1.3.repo"}

2015-06-18 15:08:14,138.138 INFO:teuthology.task.ansible.out:
TASK: [downstream-setup | Set remove_yum_repos on teardown] *******************

2015-06-18 15:08:14,156.156 INFO:teuthology.task.ansible.out:skipping: [magna085.ceph.redhat.com]

2015-06-18 15:08:14,157.157 INFO:teuthology.task.ansible.out:
TASK: [downstream-setup | Delete yum repos from /etc/yum.repos.d] *************

2015-06-18 15:08:14,169.169 INFO:teuthology.task.ansible.out:skipping: [magna085.ceph.redhat.com]

2015-06-18 15:08:14,171.171 INFO:teuthology.task.ansible.out:
PLAY RECAP ********************************************************************
magna085.ceph.redhat.com   : ok=2    changed=1    unreachable=0    failed=0


2015-06-18 15:08:14,288.288 INFO:teuthology.run_tasks:Running task interactive...
Ceph test interactive mode, use ctx to interact with the cluster, press control-D to exit...
>>>
2015-06-18 15:08:52,818.818 DEBUG:teuthology.run_tasks:Unwinding manager ansible
2015-06-18 15:08:52,818.818 DEBUG:teuthology.task.ansible:Running ansible-playbook -v --extra-vars '{"yum_repos": [{"url": "http://puddle.ceph.redhat.com/puddles/RHCeph/1.3-RHEL-7/l
atest/Server-RH7-CEPH-MON-1.3.repo", "name": "mon"}], "teardown": true, "ansible_ssh_user": "ubuntu"}' -i /etc/ansible/hosts --limit magna085.ceph.redhat.com /home/andrewschoen/src/
ceph-cm-ansible_master/downstream_setup.yml
2015-06-18 15:08:53,272.272 INFO:teuthology.task.ansible.out:
PLAY [testnodes] **************************************************************

2015-06-18 15:08:53,274.274 INFO:teuthology.task.ansible.out:
GATHERING FACTS ***************************************************************

2015-06-18 15:08:53,614.614 INFO:teuthology.task.ansible.out:ok: [magna085.ceph.redhat.com]

2015-06-18 15:08:53,616.616 INFO:teuthology.task.ansible.out:
TASK: [downstream-setup | Download yum repos to /etc/yum.repos.d] *************

2015-06-18 15:08:53,634.634 INFO:teuthology.task.ansible.out:skipping: [magna085.ceph.redhat.com] => (item={u'url': u'http://puddle.ceph.redhat.com/puddles/RHCeph/1.3-RHEL-7/latest/
Server-RH7-CEPH-MON-1.3.repo', u'name': u'mon'})

2015-06-18 15:08:53,636.636 INFO:teuthology.task.ansible.out:
TASK: [downstream-setup | Set remove_yum_repos on teardown] *******************

2015-06-18 15:08:53,653.653 INFO:teuthology.task.ansible.out:ok: [magna085.ceph.redhat.com] => {"ansible_facts": {"remove_yum_repos": ["mon"]}}

2015-06-18 15:08:53,654.654 INFO:teuthology.task.ansible.out:
TASK: [downstream-setup | Delete yum repos from /etc/yum.repos.d] *************

2015-06-18 15:08:53,794.794 INFO:teuthology.task.ansible.out:changed: [magna085.ceph.redhat.com] => (item=mon) => {"changed": true, "item": "mon", "path": "/etc/yum.repos.d/mon.repo
", "state": "absent"}

2015-06-18 15:08:53,796.796 INFO:teuthology.task.ansible.out:
PLAY RECAP ********************************************************************
magna085.ceph.redhat.com   : ok=4    changed=1    unreachable=0    failed=0
```

I used this config:

```
   interactive-on-error: true                                                                                                                                                 
   roles:
       - [mon.0, client.0]
   tasks:
       - ansible:
           repo: https://github.com/ceph/ceph-cm-ansible.git
           playbook: downstream_setup.yml
           cleanup: True
           vars:
            yum_repos:
                - url: http://puddle.ceph.redhat.com/puddles/RHCeph/1.3-RHEL-7/latest/Server-RH7-CEPH-MON-1.3.repo
                  name: mon
      - interactive:
```

zmc commented on teuthology/task/ansible.py:
```diff
@@ -250,12 +254,28 @@ def _build_args(self):
 
     def teardown(self):
+        if self.config.get("cleanup"):
+            self._cleanup()
```
This should go after the file removals

andrewschoen replied:
Would it not need them to run the playbook in some instances?

zmc commented on teuthology/task/ansible.py:
```diff
@@ -250,12 +254,28 @@ def _build_args(self):
+    def _cleanup(self):
+        """
+        If the ``cleanup`` key exists in config the same playbook will be
+        ran again during the teardown step with the var ``teardown`` given with
```
*run

zmc replied:
Also, this docstring makes it sounds like the function doesn't do anything if 'cleanup' isn't set. I think we should either clarify the docstring or make ``_cleanup()`` itself do the config check - if you don't have strong feelings, I think I prefer the latter.

andrewschoen replied:
You're proposing I always call ``_cleanup`` in ``teardown``, but then noop in ``_cleanup`` if 'cleanup' isn't in config?

Sure, I can do that.

ceph-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.ceph.com//job/teuthology-pull-requests/1528/
Tests passed for this pull request.

zmc commented on teuthology/task/ansible.py:
```diff
@@ -62,6 +62,10 @@ class Ansible(Task):
                     directly to ansible-playbook.
         vars:       A dict of vars to be passed to ansible-playbook via the
                     --extra-vars flag
+        cleanup:    If present, the given or generated playbook will be ran again
```
*run

andrewschoen replied:
Good grief, will fix.

zmc commented on teuthology/task/ansible.py:
```diff
@@ -250,12 +254,31 @@ def _build_args(self):
+                self.config['vars'] = extra
+            self.execute_playbook()
+
+        log.info("Skipping ansible cleanup...")
```
Looks like this will say it's skipping cleanup even after it runs cleanup

andrewschoen replied:
You're right, I'll fix that.

ceph-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.ceph.com//job/teuthology-pull-requests/1529/
Tests passed for this pull request.

zmc commented:
This looks great!
